src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs

```diff
@@ -577,22 +577,26 @@ private bool AcquireClientCredentials(ref byte[]? thumbPrint)
                 {
                     if (NetEventSource.Log.IsEnabled())
                         NetEventSource.Log.UsingCachedCredential(this);
-
                     _credentialsHandle = cachedCredentialHandle;
                     _selectedClientCertificate = clientCertificate;
                     cachedCred = true;
                 }
                 else
                 {
-                    _credentialsHandle = SslStreamPal.AcquireCredentialsHandle(selectedCert!, _sslAuthenticationOptions.EnabledSslProtocols, _sslAuthenticationOptions.EncryptionPolicy, _sslAuthenticationOptions.IsServer);
+                    if (selectedCert != null)
+                    {
+                        _sslAuthenticationOptions.CertificateContext = SslStreamCertificateContext.Create(selectedCert!);
+                    }
+                    _credentialsHandle = SslStreamPal.AcquireCredentialsHandle(_sslAuthenticationOptions.CertificateContext,
+                            _sslAuthenticationOptions.EnabledSslProtocols, _sslAuthenticationOptions.EncryptionPolicy, _sslAuthenticationOptions.IsServer);
 
                     thumbPrint = guessedThumbPrint; // Delay until here in case something above threw.
                     _selectedClientCertificate = clientCertificate;
                 }
             }
             finally
             {
-                if (selectedCert != null)
+                if (selectedCert != null && _sslAuthenticationOptions.CertificateContext != null)
```

Comment:
Personally, I don't understand what this finally is doing.  Might be worth a comment.

Reply:
updated. Originally, the context was only updated here. But for OSX we need it before AcquireCredentialsHandle() above is called. So this is guard to do it only once e.g. cover all the cases other than line 588.